doc/design/command-line-parser.md

```diff
@@ -0,0 +1,162 @@
+| option argument   | Argument which is separated by a space from an option, like `123` or `blubb`. |
+| argc & argv       | Taken from `int main(int argc, char* argv[])`. Provides access to the command line arguments. |
+
+### Option Types
+
+| Name              | Description                                              |
+| :---------------- | :------------------------------------------------------- |
+| switch            | When provided as argument the switch is activated, see `-h` and `--help`. A boolean signals to the user whether the switch was provided (activated) or not. |
+| optional          | A value which can be provided via command line but is not required. Every optional option requires a default value. |
+| required          | A value which has to be provided via command line. Does not require a default value. |
+
+Every option has an option argument but since the switch is boolean in nature
+the argument is implicitly provided when stating the option.
+
+## Design
+
+### Considerations
+
+The solution shall be:
+
+ * easy to use by the developer
+ * provide a unified appearance and syntax in all applications with command line arguments
+    * the help must be generated by the parser
+    * the help shall contain the type if an argument requires a value
+    * the syntax is defined by the parser not the developer as much as possible
+ * shall be type safe when command line arguments are converted to types like `int` or `float`
+    * underflow, overflow shall be handled
+    * a detailed error explanation shall be presented to the user when cast fails
+
+### Solution
+
+#### Class Diagram
+
+![class diagram](../website/images/command_line_parser_class_overview.svg)
+
+#### Sequence Diagram
+
+Lets assume we would like to add a switch, an optional and a required option, parse
+them and print them on the console.
+
+![sequence diagram](../website/images/command_line_parser_usage.svg)
+
+#### Macro Based Code Generator
+
+The macros `IOX_CLI_DEFINITION`, `IOX_CLI_SWITCH`, `IOX_CLI_OPTIONAL` and `IOX_CLI_REQUIRED`
+provide building blocks so that the user can generate a struct. The members of that
+struct are defined via the macros and set in the constructor of that struct which
+will use the `OptionManager` to parse and extract the
+values safely.
+
+![macro sequence diagram](../website/images/command_line_parser_macro_usage.svg)
+
+The struct constructor can be called with `argc` and `argv` as arguments from 
+`int main(int argc, char* argv[])`.
+
+### Code example
+
+```cpp
+struct CommandLine
+{
+    IOX_CLI_DEFINITION(CommandLine, "My program description");
+
+    IOX_CLI_OPTIONAL(string<100>, stringValue, {"default Value"}, 's', "string-value", "some description");
+    IOX_CLI_REQUIRED(string<100>, anotherString, 'a', "another-string", "some description");
+    IOX_CLI_SWITCH(doStuff, 'd', "do-stuff", "do some stuff - some description");
+    IOX_CLI_OPTIONAL(uint64_t, version, 0, 'v', "version", "some description");
```

Comment:
I would either remove this last option, give it a different name or change it to a switch. It's weird to be able to call `foo --version 42`. Wouldn't it make sense to have the version optional in `IOX_CLI_DEFINITION` and just print the version and exit the application if the flag was used?
Similar to what is done with `-h`?
e.g.
```cpp
IOX_CLI_DEFINITION(CommandLine, "My program description", 1, 0);
```
or maybe an additional
```cpp
IOX_CLI_VERSION(1, 0);
```

Reply:
On the other hand, for RouDi we also print the build date and commit hash. With the proposal from above, this won't work